### DISCUSSION TITLE
VST and permutation test

### DISCUSSION THREAD

**Discussion by JorgeCortes-M** [piyalkarum/rCNV#2] [Ideas]
Dear Piyal, this package is awesome and help me a lot in my research, thank you. I have been thinking and testing the implementation of a permutation test in the estimation of the VST index in order to test the statistical significance. I think that it could be similar to an analysis with FST. What do you think about permutation test implemented for VST?

**Comment by piyalkarum** [comment]
Hi Jorge,
Thanks for using the package. I'm glad you found it useful.
Your idea of permutation of vst is excellent. That'll provide both a significance test and a different way to present the CNV variation among populations. I think the implementation of a permutation step should be easy.  I would sub-sample the populations x number of times to get a bootstrap value. Was that your idea? I'm interested to know how you would implement it.

**Comment by JorgeCortes-M** [comment]
Thanks for your answer Piyal,
I tried a very simple test, I reassign the populations to the analyzed individuals in order to represent a random distribution, and then with the reassigned individuals I used the function vst, then I stored the results in a dataframe for each pair of population comparisons and finally compare it to the real value obtained by the vst function with the individuals in his real populations. Maybe with the code below it will be more clear.

```
nrep = 1000

df <- data.frame()

for(i in 1:nrep){
  pop_test <- sample(pop_data$pop, size = length(pop_data$pop), replace = F)
  Vs_sim <-  vst(cnv_rd_data, pops= pop_test)
  df_test <- data.frame(vst_sfm_pel = Vs_sim[2,1], vst_sfm_im = Vs_sim[3,1],
                        vst_sfm_mel = Vs_sim[4,1], vst_pel_im = Vs_sim[3,2],
                        vst_pel_mel = Vs_sim[4,2], vst_im_mel = Vs_sim[4,3])
  df <- rbind(df, df_test)
}

df
```

Thats for generate the dataframe with the resample values for four populations (sfm, pel, mel and im, in the file pop_data) and then to calculate the p-value I used the next code:

```
Vs <-  vst(cnv_rd_data, pops= pop_data$pop)

df %>%
  filter(vst_sfm_mel > Vs[4,1]) %>%
  dplyr::select(vst_sfm_mel) %>%
  count() %>%
  mutate(pvalue = n/nrep)
```

What do you think about it? the main issue about it is that when I use the vst function in every run of the for loop a new plot is generated and maybe that consume more resources than those that would be needed for the analysis.

Thanks for your time!

**Comment by piyalkarum** [comment]
Hi Jorge,
Apologies for the delayed reply.

Your script should work just fine for the permutation test. My suggestion would have been the same as your population name reassignment and comparing it to the original value. One additional step would be to plot the distribution of the vst values from the permutations as a histogram and add the original vst value as a vertical line to show the significance. I will try to implement it when I get some free time. Or if you can use our example data to implement this, feel free to send me a script. I can include it in the package and I will add you as an author for the function.

As for the plotting when you run the vst function, there is an option in the vst function to turn off the plotting. If you set "qGraph=FALSE," it will only calculate the VST values and won't produce a plot.

Let me know if this helps.
Thanks again for using the package.

**Comment by JorgeCortes-M** [comment]
Hi Piyal,

Thanks for your answer, I will try to generate the script for the function using the example data, I will attach the script when I finish it. Thanks for the option for vst function I will use it.

**Comment by JorgeCortes-M** [comment]
Hi Piyal,

I was trying to test the code in the example data (example.raw.vcf.gz), but I don't know the populations of each sample to run the vst function. Do you have that information or maybe I have to test the function with another datasets, like Dorant et al., 2020?

Thanks again!

**Comment by piyalkarum** [comment]
The population names in the example data set should be included in the first 11 letters of the sample name as it is done in the example of `vst()` function. This `substr(colnames(ad)[-c(1:4)],1,11)` should get you that. Otherwise, if you're comfortable using the Dorant data, that works too.

**Comment by JorgeCortes-M** [comment]
Thanks Piyal, I will use the example data then!

**Comment by JorgeCortes-M** [comment]
Hello Piyal,

I have been working in this function, finally I decided to permute the individuals because I noted that is best in order to mantain the abstraction for any dataset type. Below is the code (I can send you the file by other media if you want) and the application to the Dorant dataset. This function return a list with different elements, the Vst values, the dataframe with the permutation vst values and the p values. The only thing that I do not implemented was the histogram, because I still do not know the best way to implement it (In the case of dorant data woul be 45 histograms, so I do not know how to store it). What do you think about it?

```
library(rCNV)


vdata<-readRDS("./testDATA/lobster_vstDat.rds")

dd<-rCNV:::apply_pb(vdata[,-c(1:4)],2,function(x){do.call(cbind,lapply(x,function(y){sum(as.numeric(unlist(strsplit(as.character(y),","))))}))})
pops<-substr(colnames(dd),1,4)
dd<-data.frame(vdata[,1:4],dd)



vstPermutation <- function(AD, pops, nrep) {
  Vs <- vst(AD, pops, qGraph = FALSE)
  df_Vs <- data.frame(matrix(ncol=choose(length(unique(pops)), 2)))
  
  for(i in 1:nrow(Vs)){
    for(j in 1:ncol(Vs)){
      col_vs <- paste(sep = '_', rownames(Vs)[i], colnames(Vs)[j])
      Vs_value <- Vs[i,j]
      df_Vs[col_vs] <- Vs_value
    }
  }
  Vs_noNA <- df_Vs[ , colSums(is.na(df_Vs))==0]
  
  df <- data.frame(matrix(ncol=choose(length(unique(pops)), 2)))
  
  df <- setNames(df, c(names(Vs_noNA))) 
  
  
  
  for(i in 1:nrep){
    ind_cnv <- AD[-1:-4]
    ind_test <- sample(ind_cnv, size = length(ind_cnv), replace = F)
    dd_test <- cbind(AD[1:4], ind_test)
    Vs_sim <-  vst(dd_test, pops= pops, qGraph = FALSE)
    df_Vs_test <- data.frame(matrix(ncol=choose(length(unique(pops)), 2),
                                    nrow=1))
    df_Vs_test <- setNames(df_Vs_test, c(names(Vs_noNA)))
    
    for(pop1 in 1:nrow(Vs_sim)){
      for(pop2 in 1:ncol(Vs_sim)){
        col_vs <- paste(sep = '_', rownames(Vs_sim)[pop1], colnames(Vs_sim)[pop2])
        if(col_vs %in% colnames(Vs_noNA)){
          Vs_value <- Vs_sim[pop1,pop2]
          df_Vs_test[col_vs] <- Vs_value
        }else{
          next
        }
      }
    }
    
    df <- rbind(df, df_Vs_test)
  }
  
  df <- df[-1,]
  
  df_pvalues <- matrix(ncol=choose(length(unique(pops)), 2))
  df_pvalues <- setNames(df_pvalues, c(names(Vs_noNA)))
  
  for(col in colnames(df)){
    count=0
    for(value_list in df[col]){
      for(value in value_list){
        if(value > Vs_noNA[col] ){
          count = count + 1
        }
      }
    }
    p_value <- count/nrep
    df_pvalues[col] <- p_value
    
  }
  
  perm_list <- list(Vs_noNA, df, df_pvalues)
  
  
  return(perm_list)
}

dd_perm <- vstPermutation(AD = dd, pops = pops, nrep = 100)
```

**Reply by piyalkarum** [reply]
Thanks for putting your time into this. Your script looks quite comprehensive. I'll review it over the holidays and try to add the histogram. Once I have an optimized function, I'll send it to you for review.

Happy Holidays!

**Reply by JorgeCortes-M** [reply]
Thanks Piyal, Happy Holidays for you too!

**Reply by piyalkarum** [reply]
Hi Jorge,
Finally, I had some time to sit down and check the code. Now the vstPermutation function is there in the GitHub package. Please check it out. I also added a histogram to show the distribution of the permutation and the observed values. Your name is also there in the author list. If you send me your email address, I can include it there in the next push. 

Please let me know if it works as it should.
Thanks again for your feedback. 

Piyal

**Reply by JorgeCortes-M** [reply]
Hi Piyal,
Thank you for improve and incorporate the function in the package. I tested it with the example dataset and with my own data, it works very well, the only thing that could be a problem is the histogram. I noted in first place that only takes in consideration one of the comparisons between populations, it is stated in the definition of the abline: `abline(v = obs_vst[2,1], col = 2, lwd = 2)`, this line only considered the compatison in obs_vst[2,1], maybe that histogram could considered variables in order to select the histogram of the comparison to be displayed, or something like that. The second thing with the histogram is that the limits in the X axis in some cases are smaller than the value of the observed vst, so the abline is not displayed in the graph, a possible solution is considered that observed value to set the X axis limits. Those are my only observations about the behaviour of the function, but it works nice and is faster than the function that I wrote in the past months.

Thank you again Piyal for taking in consideration my suggestions. Also my email is: jorge.cortes.m@ug.uchile.cl
I hope this feedback could be useful!

Jorge

**Reply by piyalkarum** [reply]
You're right about the histogram using only one vst permutation. That was an oversight from my side. Now, I changed it to plot either the mean absolute distance or the square root of the squared distance, and also provide the user the option to choose either. Therefore, essentially, now it takes all the vst permutations to plot the distribution. I also fixed the x-axis range limitation. Please check it and let me know if it works fine. Thanks for the feedback.

**Reply by JorgeCortes-M** [reply]
Piyal, It works perfectly now. Thank you for add this function to the package. I've been using the package to analyze my data so in case that you need to test something or if a new idea come to me, we could be in contact!

**Reply by piyalkarum** [reply]
Wonderful!
Thanks again for using the package and the very useful input. Please feel free to reach out if you have any other ideas/suggestions and I'd do the same.
